packages/dd-trace/src/profiling/profilers/wall.js

```diff
@@ -142,42 +149,130 @@ class NativeWallProfiler {
+        // with an empty object.
+
+        // Cache lookup of web context for this span context. It's either itself or one of its
+        // parent spans' contexts.
+        const webContext = spanContext[WebContextSymbol]
+        if (webContext === undefined) {
+          // Not cached yet. Find the first webspan starting from the innermost span. There might be
+          // several webspans, for example with next.js, http plugin creates a first span and then
+          // next.js plugin creates a child span, and this child span has the correct endpoint
+          // information.
+          let found = false
+          for (let i = startedSpans.length - 1; i >= 0; i--) {
+            const sspan = startedSpans[i]
+            if (sspan._duration !== undefined) {
+              // span is listed in trace started, but it finished in the meantime
+              continue
+            }
+            const scontext = startedSpans[i].context()
+            const tags = scontext._tags
+            if (isWebServerSpan(tags)) {
+              this._lastWebContext = scontext
+              this._lastWebTags = tags
+              // Cache so next time this span activates we don't need to lookup again
+              spanContext[WebContextSymbol] = scontext
+              found = true
+              break
+            }
+          }
+          if (!found) {
+            this._lastWebContext = undefined
+            this._lastWebTags = undefined
+            // Cache negative lookup so next time this span activates we don't need to lookup again
+            spanContext[WebContextSymbol] = null
+          }
+        } else if (webContext === null) {
+          // Use negative cache lookup info
+          this._lastWebTags = undefined
+          this._lastWebContext = undefined
+        } else {
+          // Use cached lookup info
+          this._lastWebContext = webContext
+          this._lastWebTags = webContext._tags
+        }
+      }
     } else {
-      this._lastStartedSpans = undefined
-      this._lastSpan = undefined
+      this._clearLast()
     }
   }
 
+  _clearLast () {
+    this._lastSpan = undefined
+    this._lastRootSpan = undefined
+    this._lastWebTags = undefined
+    this._lastWebContext = undefined
+  }
+
   _updateContext (context) {
-    if (!this._lastSpan) {
-      return
-    }
-    if (this._codeHotspotsEnabled) {
+    if (this._lastSpan) {
       context.spanId = this._lastSpan.context().toSpanId()
-      const rootSpan = this._lastStartedSpans[0]
-      if (rootSpan) {
-        context.rootSpanId = rootSpan.context().toSpanId()
+    }
+    if (this._lastRootSpan) {
+      context.rootSpanId = this._lastRootSpan.context().toSpanId()
+    }
+    if (this._lastWebTags) {
+      // Try to eagerly compute the endpoint name here
+      const endpoint = endpointNameFromTags(this._lastWebTags)
+      if (endpoint) {
+        context.endpoint = endpoint
+      } else {
```

Comment:
I'm actually curious when can we enter this branch. We'd enter it if at the time of sampling, the span didn't already have the tags needed for endpoint computation. When we originally merged endpoint tracing, we added an `addRequestTags` to `startSpan()` in addition to `finishSpan()` in `web.js`, see https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/util/web.js#L106. I'm curious if we have any remaining cases when the request tags – used for establishing the endpoint – would not be present early in the span's lifecycle.